### PR TITLE
add whole event log callback module

### PR DESCRIPTION
**Pull request by ywy2090**

### CODE REVIEW

File: libevent/EventLogFilter.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+bool EventLogFilter::matches(LogEntry const& _log)
+{
+    auto addresses = getParams()->getAddresses();
+    auto topics = getParams()->getTopics();
+    // An empty address array matches all values otherwise log.address must be in addresses
+    if (!addresses.empty() && !addresses.count(_log.address))
+        return false;
+
+    bool isMatch = true;
+    for (unsigned i = 0; i < 4; ++i)
```
Comment (morebtcg) [inline libevent/EventLogFilter.cpp]: use constant

File: libevent/EventLogFilterParams.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    // m_groupID
+    int getGroupID() const { return m_groupID; }
+    // m_startBlock
+    eth::BlockNumber getStartBlock() const { return m_startBlock; }
+    // m_endBlock
+    eth::BlockNumber getEndBlock() const { return m_endBlock; }
+    // m_addresses
+    const AddressHash& getAddresses() const { return m_addresses; }
+    // m_topics
+    const std::array<h256Hash, 4>& getTopics() const { return m_topics; }
```
Comment (morebtcg) [inline libevent/EventLogFilterParams.h]: use constant

File: libevent/EventLogFilterManager.h [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+using EventLogFilterVector = std::vector<EventLogFilter::Ptr>;
+
+class EventLogFilterManager : public Worker
+{
+public:
+    using Ptr = std::shared_ptr<EventLogFilterManager>;
+
+public:
+    // constructor function
+    EventLogFilterManager(std::shared_ptr<dev::ledger::LedgerManager> _ledgerManager,
```
Comment (morebtcg) [inline libevent/EventLogFilterManager.h]: inside the ledeger, triger by rpc & blockchain

File: libevent/EventLogFilterManager.h [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        int64_t _maxBlockRange = -1, int64_t _maxBlockPerLoopFilter = 10)
+      : m_ledgerManager(_ledgerManager),
+        m_maxBlockRange(_maxBlockRange),
```
Comment (morebtcg) [inline libevent/EventLogFilterManager.h]: config.ini

File: libevent/EventLogFilterManager.h [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    bool isErrorStatus(filter_status status)
+    {
+        return ((status == filter_status::GROUP_ID_NOT_EXIST) ||
+                (status == filter_status::CALLBACK_FAILED) ||
+                (status == filter_status::ERROR_STATUS));
+    }
+
+    // main loop thread
+    void doWork() override;
+    void doAddFilter();
```
Comment (morebtcg) [inline libevent/EventLogFilterManager.h]:
```suggestion
    void addFilter();
```

File: libevent/EventLogFilterManager.h [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    void doFilter();
+    filter_status doFilter(EventLogFilter::Ptr _filter);
```
Comment (morebtcg) [inline libevent/EventLogFilterManager.h]:
```suggestion
    filter_status executeFilter(EventLogFilter::Ptr _filter);
```

File: libchannelserver/ChannelRPCServer.cpp [hunk 7]
```diff
@@ -372,6 +376,59 @@ void dev::ChannelRPCServer::onClientRPCRequest(
...
+        auto sessionRef = std::weak_ptr<dev::channel::ChannelSession>(session);
+        auto serverRef = std::weak_ptr<dev::channel::ChannelServer>(_server);
+        auto protocolVersion = static_cast<uint32_t>(session->protocolVersion());
+
+        if (m_eventFilterCallBack)
+        {
+            m_eventFilterCallBack(
+                data,
+                [serverRef, sessionRef, seq](
+                    int32_t retCode, const std::string& response, bool isSend) {
```
[the remark this answers is not in the document]
Comment (ywy2090) [reply libchannelserver/ChannelRPCServer.cpp]: done

> use shouldSend insted of isSend

done

File: libchannelserver/ChannelRPCServer.cpp [hunk 8]
```diff
@@ -372,6 +376,59 @@ void dev::ChannelRPCServer::onClientRPCRequest(
...
+                    auto server = serverRef.lock();
+                    auto session = sessionRef.lock();
+
+                    if (server && session && session->actived())
+                    {
+                        if (isSend)
+                        {
+                            auto channelMessage = server->messageFactory()->buildMessage();
+                            channelMessage->setType(0x15);
```
[the remark this answers is not in the document]
Comment (ywy2090) [reply libchannelserver/ChannelRPCServer.cpp]: > use another protocol type

done

File: libchannelserver/ChannelRPCServer.cpp [hunk 9]
```diff
@@ -372,6 +376,59 @@ void dev::ChannelRPCServer::onClientRPCRequest(
...
+                                std::function<void(dev::channel::ChannelException, Message::Ptr)>(),
+                                0);
+                        }
+
+                        return true;
+                    }
+
+                    return false;
+                },
+                [protocolVersion]() -> uint32_t { return protocolVersion; });
```
[the remark this answers is not in the document]
Comment (ywy2090) [reply libchannelserver/ChannelRPCServer.cpp]: > store protocolVersion in Task member

done

File: libevent/Common.h [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+namespace dev
+{
+namespace event
+{
+using MatchedLogEntry = std::pair<eth::LogEntry, eth::TransactionReceipt>;
+using MatchedLogEntries = std::vector<MatchedLogEntry>;
+
+enum class ResponseCode
+{
+    SUCCESS = 0,
```
[the remark this answers is not in the document]
Comment (ywy2090) [reply libevent/Common.h]: > add 1 desc

done

File: libevent/EventLogFilter.cpp (hunk 1 above)
Comment (ywy2090) [reply libevent/EventLogFilter.cpp]: > use constant

done

File: libevent/EventLogFilterManager.h (hunk 5 above)
Comment (ywy2090) [reply libevent/EventLogFilterManager.h]: > 

done

File: libinitializer/EventLogFilterManagerInitializer.cpp [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+{
+    int64_t maxBlockRange = _pt.get<int64_t>("filter.block_range", 1000);
+    int64_t maxBlockPerProcess = _pt.get<int64_t>("filter.max_block_per_process", 10);
+
+    INITIALIZER_LOG(DEBUG) << LOG_BADGE("EventLogFilterManagerInitializer")
+                           << LOG_DESC("initConfig") << LOG_KV("maxBlockRange", maxBlockRange)
+                           << LOG_KV("maxBlockPerProcess", maxBlockPerProcess);
+
+
+    m_eventLogFilterManager =
```
Comment (wheatli) [inline libinitializer/EventLogFilterManagerInitializer.cpp]: You'd better give a setEventLogFilterManager for testing

File: libevent/EventLogFilterParams.h [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    // Add a matching address to m_addresses
+    void addAddress(Address const& _a) { m_addresses.insert(_a); }
+    // Add a matching topic to m_addresses
+    bool addTopic(unsigned _index, h256 const& _t)
+    {
+        if (_index < 4)
+        {
+            m_topics[_index].insert(_t);
```
Comment (wheatli) [inline libevent/EventLogFilterParams.h]: overwrite the array?

File: libevent/EventLogFilterParams.h [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+        return (_index < 4);
+    }
+
+private:
+    int m_groupID;
+    eth::BlockNumber m_startBlock;
+    eth::BlockNumber m_endBlock;
+    AddressHash m_addresses;
+    std::array<h256Hash, 4> m_topics;
```
Comment (wheatli) [inline libevent/EventLogFilterParams.h]: what's the meaning of 4?

File: libevent/EventLogFilterParams.cpp [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+            Json::Value root;
+
+            if (!reader.parse(_json, root) || !root.isObject())
+            {  // json parser failed
+                strDesc = " json parser failed.";
+                break;
+            }
+
+            // groupID field
+            if (!root.isMember("groupId"))
```
Comment (wheatli) [inline libevent/EventLogFilterParams.cpp]: Pack all these checking logic about JSON members into a single function

File: libevent/EventLogFilterParams.cpp [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+
+            if (root["topics"].isArray())
+            {  // Multiple topics
+                if (root["topics"].size() > 4)
+                {  // The log contains at most four topics
+                    strDesc = "topics field array size more than 4.";
+                    break;
+                }
+
+                for (Json::ArrayIndex i = 0; i < root["topics"].size(); i++)
```
Comment (wheatli) [inline libevent/EventLogFilterParams.cpp]: encapsulate this logic to a function

**Pull request closed without merging**